### PR TITLE
don't call `BlobVec::grow_exact` for zero-sized layouts

### PR DESCRIPTION
**Pull request by jakobhellermann**
# Objective

Calling `std::alloc::alloc` with a zero-sized layout is UB: https://doc.rust-lang.org/stable/std/alloc/trait.GlobalAlloc.html#safety-1
Currently, when creating a zero-sized `BlobVec` and `reserve_exact` is called with an additional amount greater than `capacity - len` (so for ZSTs `usize::MAX - len`), `std::alloc::realloc` would get called.

This is a pretty rare edge case, so while it's worth fixing I don't think this would happen anywhere in real world bevy code.

## Solution

Don't call `grow_exact` if the item layout size is zero.

### CODE REVIEW

File: crates/bevy_ecs/src/storage/blob_vec.rs [hunk 1]
```diff
@@ -83,33 +83,32 @@ impl BlobVec {
 
     pub fn reserve_exact(&mut self, additional: usize) {
         let available_space = self.capacity - self.len;
-        if available_space < additional {
-            self.grow_exact(additional - available_space);
+        if available_space < additional && self.item_layout.size() > 0 {
```
Comment (bjorn3) [inline crates/bevy_ecs/src/storage/blob_vec.rs:86]: If the item size is 0, the capacity is set to `usize::MAX` and thus `available_space` would always be at least as large as `additional` and thus `grow_exact` would never be called.

File: crates/bevy_ecs/src/storage/blob_vec.rs [hunk 2]
```diff
@@ -83,33 +83,32 @@ impl BlobVec {
...
-
-            self.data = NonNull::new(new_data).unwrap_or_else(|| handle_alloc_error(new_layout));
-        }
+        let new_data = if self.capacity == 0 {
+            std::alloc::alloc(new_layout)
+        } else {
+            std::alloc::realloc(
+                self.get_ptr_mut().as_ptr(),
+                array_layout(&self.item_layout, self.capacity)
+                    .expect("array layout should be valid"),
```
Comment (alice-i-cecile) [inline crates/bevy_ecs/src/storage/blob_vec.rs:106]: This is not really a very helpful error message. I'd probably prefer an unwrap.

File: crates/bevy_ecs/src/storage/blob_vec.rs (hunk 1 above)
Comment (bjorn3) [reply crates/bevy_ecs/src/storage/blob_vec.rs:86]: Maybe `grow_exact` should take the desired capacity? Then we would do `let new_capacity = self.len + additional;` and `self.capacity < new_capacity` would can't ever return true for ZST's even if `self.len` is non-zero as `self.len + additional` would overflow when reserving `usize::MAX` elements.

Comment (alice-i-cecile) [reply crates/bevy_ecs/src/storage/blob_vec.rs:86]: I think that API would be significantly easier to use correctly as well.

**Comment by james7132** [comment]
This has been reimplemented into #4835.

**Pull request closed without merging**